[WIP] Dbt readme (#6)
bcgov/bcstats-duckdb-storage-mvp: sei_dbt/README.md

Reviewer: The bottom section of this file still needs some updates: the `Usage` section should again be specific to the project. Is there a goal in mind? If so, what files need to be run, in which order, to accomplish it?

Author: I agree, but this minimum example serves as a tool for DBAs to get started. The source code will be much easier to read.

## Overview

This project is created to load multiple CSV files into a DuckDB database using **dbt-duckdb** (Data Build Tool). It helps automate the process of transforming, modeling, and querying data from CSV files in a simple and efficient way. DuckDB is used as the local database engine, and dbt orchestrates the loading and transformation process.

> ## Rationale

Reviewer: can drop the ===== below rationale here - markdown isn't doing anything with it.

Author: Delete it.

## Overview of dbt-duckdb

This project is designed to load multiple CSV files into a DuckDB database using **dbt** (Data Build Tool). It helps automate the process of transforming, modeling, and querying data from CSV files in a simple and efficient way. DuckDB is used as the local database engine, and dbt orchestrates the loading and transformation process.

## Features

=======

> - **Automated CSV Loading:** Easily load multiple CSV files into DuckDB.

Reviewer: These 'features' are features of DTB, not really this repo. Probably don't need here.

Author: Right. It is weird to call it 'feature'. It is more like a readme for dbt-duckdb. Will change it for our project.

- **Data Transformation:** Use dbt to perform SQL-based data transformations on the loaded CSVs.
- **Lightweight Database:** Leverages DuckDB, which is optimized for efficient data processing on local files.
- **Flexible:** Customize dbt models to suit your business logic and transformation requirements.

---

## Requirements

Since we use `dbt-duckdb` Python package, before starting, ensure you have the following tools installed:

1. **Python** (>= 3.7)
2. **dbt** (>= 0.19.0)
3. **DuckDB** (Python package or standalone)
4. **CSV Files:** Ensure you have your source CSV files ready.

To install the required Python packages:

```bash

> pip install dbt-duckdb duckdb

Reviewer: My preference is to use environments, so I would probably create a conda environment from a command line, and then install there and access from the command line with that conda environment active. It means a couple of extra steps everytime you want to use dbt, but keeps your python environments clean.

Author: add this approach.

```

More instructions are available in https://github.com/mehd-io/dbt-duckdb-tutorial/

---

## Setup

Follow these steps to set up and configure the project:

1. **Clone the repository:**

   ```bash
   git clone <your-repository-url>
   cd <project-folder>
   ```

> 2. **Initiate a new projec:**

Reviewer: ```suggestion
2. **Initiate a new project:**
```

Author: Thanks for catching it. Changed it.

If you start a new project, 

```bash
dbt init
```
It will:

    ask you to name your project
    ask you which database adapter you're using (or to Supported Data Platforms)
    prompt you for each piece of information that dbt needs to connect to that database: things like account, user, password, etc

Then, it will:

    Create a new folder with your project name and sample files, enough to get you started with dbt
    Create a connection profile on your local machine. The default location is ~/.dbt/profiles.yml.

> 

Reviewer: I would add a note here that for this repo, that has already been done - file structure has been set up in `sei_dbt`

Author: Add explanation for it.

3. **Configure dbt profiles:**

>    dbt uses a `profiles.yml` file to connect to DuckDB. Ensure that your `profiles.yml` file is correctly configured as follows:

Reviewer: does this profile file have to be pushed to github? can it live locally only, and included in gitignore? (I'm not sure if it's actually needed on github or not, I am naively learning about dbt here)

Author: This is a great question. I researched it and it turns out that the profile file could be used to save credentials and it should be git ignored. The profile file could be global instead of project-specific. I will do more research and come back this point. The safepath problem could be solved using the profile file.

Author: Yes. Now `profiles.yml` is located in user home directory.

   ```yaml
   duckdb_project:
     target: dev
     outputs:
       dev:
         type: duckdb
         path: "path/to/your/database.db"  # Path to your DuckDB database file
         extensions: ['parquet', 'json']   # Optional extensions to load, if needed
            duckdb_project:

       test:
         type: duckdb
         path: "test.db"  # Test database path
         
       prod:
         type: duckdb
         path: "prod.db"  # Production database path
   ```

> 4. **Set up the directory structure:**

Reviewer: Is this step needed? The directory structure is already set up.

Reviewer: Perhaps it's not instructions to set up the strucutre, but just explaining what the structure is.

Author: Right. It just explains why we have those folders. Deleting those unused folders might be a good idea. We can get them back when we use them.

   Your dbt project should follow this structure:

   ```
   .
   ├── models
   │   ├── staging
   │   │   └── csv_imports.sql  # SQL to define CSV imports
   │   ├── transforms.sql       # SQL transformation models
   |   ├── source.yml  
   ├── data
   │   └── source.csv           # Place your CSV files here
   ├── dbt_project.yml          # dbt project configuration
   └── README.md                # This file
   ```

---

## Usage

### Loading CSV Files

1. **Prepare your CSV files:**
   - Place the source CSV files inside the `data/` directory.

2. **Create staging models:**
   - In the `models/staging` folder, create a model (SQL file) that defines how each CSV should be loaded into DuckDB.

   Example SQL to load a CSV:

   ```sql
   {{ config(materialized='table') }}

   select * 
   from read_csv_auto('data/source.csv');
   ```

3. **Run dbt to load and transform data:**

   Once your staging models are ready, run dbt to load the CSVs into DuckDB:

>    ```bash

Reviewer: run how?

Author: add more explanation for it.

   git clone <your-repository-url>
   cd <project-folder>
   ```
   
2. **Install the required Python packages**

After you clone the repo, make sure there is `requirment.txt` file.

In this project, we recommend using a **Conda environment** to manage dependencies and keep your Python environments clean. 

To get started, create a new Conda environment from the command line with the following command: `conda create -n <env_name> python=<version>`. 

Once the environment is created, activate it with `conda activate <env_name>`. 

You can then install all necessary dependencies within this environment using

> ```bash

Reviewer: It is not best practice to mix conda and pip together - we can address this later on however.

Author: Right. Switch to conda install.

pip install -r requirements.txt
```

If you already have a python environment to work with, you can

```bash
pip install dbt-duckdb duckdb
```

Every time you work on the project, ensure the environment is active by running `conda activate <env_name>` to access the correct setup. 

This approach isolates dependencies, minimizing conflicts across projects.
   
   

> 2. **Initiate a new project:**

Reviewer: ```suggestion
3. **Initiate a new project:**
```

         path: "path/to/your/database.db"  # Path to your DuckDB database file
         extensions: ['parquet', 'json']   # Optional extensions to load, if needed
            duckdb_project:
               threads: 4
         external_root: "data/dev"

       test:
         type: duckdb
         path: "test.db"  # Test database path
         external_root: "data/test" 
         
       prod:
         type: duckdb
         path: "prod.db"  # Production database path
         external_root: "data/prod"
   ```
   

> In this `profiles.yml`:

Reviewer: This section should be written less generally, and more with the specific use-case of this repo in mind. What type of profiles.yml is needed for someone working with this repo? Do they really need test/prod/dev database paths? Where should they store it?

Author: Agree. Will make the changes.

- We have three environments: `dev`, `test` and `prod`.
- For each environment, we define:
  - `external_root`: A variable points to the directory where CSV files or other data are located.  
  - `dbt-duckdb` only support certain keys in the `profiles.yml`; for example, `external_root`, `settings`, and `config_options`, etc, and each has its own purpose. `custom`, `vars` are not supported. 
   
#### Key Usage:

> - **Global Connection Configuration**: You use `profiles.yml` to store environment-specific details such as database credentials, which will be applied to all dbt projects that use this profile.

Reviewer: Again, this is more of a high level commentary than it is related to the actual code of this project. I would cut down/eliminate the key usage section, and describe how exactly the `profile.yml` file is used/created for this particular project.

Author: Agree. Add specific information for this project.

- **Sensitive Information**: The `profiles.yml` file is the right place to store sensitive information like database credentials, API tokens, or connection settings. It is generally **not version-controlled**, as it may contain secrets.
- **Multiple Environments**: You can define multiple environments (like `dev`, `prod`, `test`) within `profiles.yml` and switch between them using the `target` option. This is useful when different teams or environments need to connect to different databases or use different credentials.

#### Default Location:
- The `profiles.yml` file is located by default in `~/.dbt/profiles.yml`, making it global across multiple dbt projects on your local machine. If you use a Windows machine, it could be in `C:\Users\myusername\.dbt`.

> - You can also override this by explicitly specifying the location of `profiles.yml` with the `DBT_PROFILES_DIR` environment variable.

Reviewer: In this section, give your recommendation for how it should be set up for this project. Should the profile file be project specific, or should it contain the info for every database you might be accessing stored in a central location? If project specific, how do you make sure that this project is looking at the right yml file?

Reviewer: https://docs.getdbt.com/docs/core/connect-data-platform/connection-profiles
from the source you reference, it sounds like it searches the current directory first before going to ~/.dbt/ - so I would suggest defaulting to project specific yaml files like it does there. Then just using `.gitignore` to make sure that the path info in that profile.yml file isn't pushed.

Author: > https://docs.getdbt.com/docs/core/connect-data-platform/connection-profiles from the source you reference, it sounds like it searches the current directory first before going to ~/.dbt/ - so I would suggest defaulting to project specific yaml files like it does there. Then just using `.gitignore` to make sure that the path info in that profile.yml file isn't pushed.

Right. The local `profiles.yml` will override the global `profiles.yml`. Either way works for us. A local project-specific `profiles.yml` is better.

To **infer variables from `profiles.yml`** in your **dbt model file**, you need to use **Jinja** templating and the `target` object, which exposes information from `profiles.yml`.

> #### Step 2: Access the Variables in the Model File

Reviewer: We went from numbered list to step 2 here, not sure how that happened.

Author: Yes. The first step is removed, so we will remove the second step. Just keep it as one step.

Here’s an example of how you might use the `data_path` variable inside a dbt model:

```sql
-- models/staging/my_staging_model.sql

WITH raw_data AS (

    -- Use the data_path from profiles.yml
    SELECT *
    FROM read_csv_auto('{{ target.external_root }}/my_file.csv')

),

final_data AS (

    SELECT
        *
    FROM raw_data

)

SELECT * FROM final_data;
```

> - **`{{ target.vars.data_path }}`**: This Jinja expression accesses the `data_path` variable from `profiles.yml`. When you run dbt with the `--target dev` flag, it will dynamically insert `"data/dev"`, and for `--target prod`, it will insert `"data/prod"`.

Reviewer: This example does not align with the example just above. 
```suggestion
- **`{{ target.external_root }}`**: This Jinja expression accesses the `external_root` variable from `profiles.yml`. When you run dbt with the `--target dev` flag, it will dynamically insert `"data/dev"`, and for `--target prod`, it will insert `"data/prod"`.
```

Author: I have changed it.

> In this case, the model will:

Reviewer: I don't think this line is necessary.

Author: I have deleted it. Thanks for catching this error.

1. **Use the `data_path`** defined in `profiles.yml` to load the appropriate CSV file based on the environment (e.g., `data/dev/my_file.csv` in dev, `data/prod/my_file.csv` in prod).

#### Source:
- [Profiles.yml Reference](https://docs.getdbt.com/docs/configure-your-profile)
   
   
4. **Configure project:**   Project-Specific Configuration

The `dbt_project.yml` file is the configuration file specific to a **single dbt project**. It contains settings related to that particular project, such as the project's name, paths to models, seeds, and specific **variables** (used in transformations and models).

This file is usually placed in the root of the dbt project directory and is included in version control because it’s tied to the project itself.

#### Example of `dbt_project.yml`:
```yaml
name: 'my_dbt_project'
version: '1.0.0'
profile: 'my_profile'

# Define the structure and directories for models
model-paths: ["models"]
seed-paths: ["seeds"]

# Project-specific variables
vars:
  some_variable: 'value'
  environment: 'dev'  # Example environment variable
```

> #### Key Usage:

Reviewer: Again, this doesn't read as project specific information, but more of a guideline for dbt in general. The README really needs to focus on how to use the core project tools, and just link to resources about learning about them elsewhere if necessary. The key usage portion seems less relevant to this project.

Author: I have changed it.

- **Variables**: You can define project-specific variables under `vars`. These are useful when you need to reference configuration values in multiple places in your models, macros, or tests.
  - These variables are **project-specific** and are **not meant to store sensitive information** like database credentials.
- **Project Paths**: You can configure paths for models, seeds, and snapshots.
- **Project-specific Settings**: This file defines what dbt will do within the scope of your project. The configurations are highly tailored to that particular project, making it version-controlled and tied to the specific project context.

#### Source:
- [dbt_project.yml Reference](https://docs.getdbt.com/reference/project-configs/dbt_project.yml)
   
   

> ### **Comparison of `dbt_project.yml` vs. `profiles.yml`**

Reviewer: This table and the best practices section that follows seems overly complicated for the project. Try to focus the README on project specific information. If we want to store our learnings about best practices for how to use dbt, that should probably be stored with the rest of the project files for now.

Author: Agree. Delete it.
